pkg/controllers/cronfederatedhpa/cronfederatedhpa_controller.go

```diff
@@ -0,0 +1,204 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	autoscalingv1alpha1 "github.com/karmada-io/karmada/pkg/apis/autoscaling/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+const (
+	// ControllerName is the controller name that will be used when reporting events.
+	ControllerName = "cronfederatedhpa-controller"
+)
+
+// CronFHPAController is used to operate CronFederatedHPA.
+type CronFHPAController struct {
+	client.Client // used to operate Cron resources.
+	EventRecorder record.EventRecorder
+
+	RateLimiterOptions ratelimiterflag.Options
+	CronHandler        *CronHandler
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *CronFHPAController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling CronFederatedHPA %s", req.NamespacedName)
+
+	cronFHPA := &autoscalingv1alpha1.CronFederatedHPA{}
+	if err := c.Client.Get(ctx, req.NamespacedName, cronFHPA); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof("Begin to cleanup the cron jobs for CronFederatedHPA:%s", req.NamespacedName)
+			c.CronHandler.StopCronFHPAExecutor(req.NamespacedName.String())
+			return controllerruntime.Result{}, nil
+		}
+
+		klog.Error("Fail to get CronFederatedHPA(%s):", req.NamespacedName, err)
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			c.EventRecorder.Event(cronFHPA, corev1.EventTypeWarning, "ProcessCronFederatedHPAFailed", err.Error())
+		}
+	}()
+
+	//  If this CronFederatedHPA is deleting, stop all related cron executors
+	if !cronFHPA.DeletionTimestamp.IsZero() {
+		c.CronHandler.StopCronFHPAExecutor(req.NamespacedName.String())
+		return controllerruntime.Result{}, nil
+	}
+
+	origRuleSets := sets.New[string]()
+	for _, history := range cronFHPA.Status.ExecutionHistories {
+		origRuleSets.Insert(history.RuleName)
+	}
+
+	// If scale target is updated, stop all the rule executors, and next steps will create the new executors
+	if c.CronHandler.CronFHPAScaleTargetRefUpdates(req.NamespacedName.String(), cronFHPA.Spec.ScaleTargetRef) {
+		c.CronHandler.StopCronFHPAExecutor(req.NamespacedName.String())
+	}
+
+	c.CronHandler.AddCronExecutorIfNotExist(req.NamespacedName.String())
+
+	newRuleSets := sets.New[string]()
+	for _, rule := range cronFHPA.Spec.Rules {
+		if err = c.processCronRule(cronFHPA, rule); err != nil {
+			return controllerruntime.Result{}, err
```

Comment:
What's the difference between `return controllerruntime.Result{Requeue: true}, err` and `return controllerruntime.Result{}, err`?  
Without `Requeue: true`, does this will retry?

Reply:
It will retry if `controllerruntime.Result{Requeue: true}` or `err` is not nil, any one of them will trigger to retry.
So without `Requeue: true`, it will retry.